### PR TITLE
new schema for advanced options

### PR DESCRIPTION
**Pull request by areed**

### CODE REVIEW

File: web/src/installers/index.ts [hunk 1]
```diff
@@ -66,388 +180,298 @@ export interface InstallerObject {
...
     }
-    if (this.registry && this.registry.version) {
-      h.update(`registry_version=${this.registry.version}`);
+    if (this.spec.contour && this.spec.contour.version) {
+      h.update(`contour_version=${this.spec.contour.version}`);
     }
-    if (this.prometheus && this.prometheus.version) {
-      h.update(`prometheus_version=${this.prometheus.version}`);
+    if (this.spec.registry && this.spec.registry.version) {
+      h.update(`registry_version=${this.spec.registry.version}`);
```
Comment (emosbaugh) [inline web/src/installers/index.ts]: looks like you reverted some of my prometheus stuff

File: web/src/client/index_test.ts [hunk 2]
```diff
@@ -314,21 +341,41 @@ describe("GET /<installerID>", () => {
...
+  });
 });
 
 describe("GET /<installerID>/join.sh", () => {
   describe("/latest/join.sh", () => {
-    const latest = Installer.latest();
+    const latest = Installer.latest().resolve();
 
-    it(`injects k8s ${latest.kubernetesVersion()}, weave ${latest.weaveVersion()}, rook ${latest.rookVersion()}, contour ${latest.contourVersion()}, registry ${latest.registryVersion()}, prometheus ${latest.prometheusVersion()}`, async () => {
+    it(`injects k8s ${latest.spec.kubernetes.version}, weave ${latest.spec.weave!.version}, rook ${latest.spec.rook!.version}, contour ${latest.spec.contour!.version}`, async () => {
```
Comment (emosbaugh) [inline web/src/client/index_test.ts]: missing prometheus

File: web/src/installers/index.ts [hunk 3]
```diff
@@ -66,388 +180,298 @@ export interface InstallerObject {
...
+      delete i.spec.weave;
+    }
+    if (!_.get(i.spec, "rook.version")) {
+      delete i.spec.rook;
+    }
+    if (!_.get(i.spec, "contour.version")) {
+      delete i.spec.contour;
+    }
+    if (!_.get(i.spec, "registry.version")) {
+      delete i.spec.registry;
```
Comment (emosbaugh) [inline web/src/installers/index.ts]: missing prometheus

File: web/src/installers/index.ts [hunk 4]
```diff
@@ -66,388 +180,298 @@ export interface InstallerObject {
...
-    if (this.prometheusVersion() && !Installer.resolvePrometheusVersion(this.prometheusVersion())) {
-      return { error: { message: `Prometheus version "${_.escape(this.prometheusVersion())}" is not supported` } };
+    if (this.spec.contour && !Installer.hasVersion("contour", this.spec.contour.version)) {
+      return { error: { message: `Contour version "${_.escape(this.spec.contour.version)}" is not supported` } };
     }
-
-    if (this.kotsadmVersion() && !Installer.resolveKotsadmVersion(this.kotsadmVersion())) {
-      return { error: { message: `Kotsadm version "${_.escape(this.kotsadmVersion())}" is not supported` } };
+    if (this.spec.registry && !Installer.hasVersion("registry", this.spec.registry.version)) {
+      return { error: { message: `Registry version "${_.escape(this.spec.registry.version)}" is not supported` } };
```
Comment (emosbaugh) [inline web/src/installers/index.ts]: missing prometheus

File: web/src/installers/index.ts [hunk 5]
```diff
@@ -66,388 +180,298 @@ export interface InstallerObject {
...
-      this.contourVersion() === "latest" &&
-      this.registryVersion() === "latest" &&
-      this.prometheusVersion() === "latest";
+    return _.isEqual(this.spec, {
+      kubernetes: { version: "latest" },
+      docker: { version: "latest" },
+      weave: { version: "latest" },
+      rook: { version: "latest" },
+      contour: { version: "latest" },
+      registry: { version: "latest" },
```
Comment (emosbaugh) [inline web/src/installers/index.ts]: missing prometheus

File: web/src/test/installers.ts [hunk 6]
```diff
@@ -3,7 +3,35 @@ import {expect} from "chai";
...
+    encryptNetwork: true
+    IPAllocRange: 10.32.0.0/12
+  contour:
+    version: latest
+  rook:
+    version: latest
+    storageClass: default
+    cephPoolReplicas: 1
+  registry:
+    version: latest
```
Comment (emosbaugh) [inline web/src/test/installers.ts]: missing prometheus

File: web/src/installers/index.ts (hunk 1 above)
Comment (areed) [reply web/src/installers/index.ts]: Prometheus is included in the non-legacy hash logic below

File: web/src/installers/index.ts [hunk 7]
```diff
@@ -66,388 +180,298 @@ export interface InstallerObject {
...
+    if (this.spec.kotsadm && this.spec.kotsadm.applicationSlug) {
+      h.update(`kotsadm_applicationSlug=${this.spec.kotsadm.applicationSlug}`);
     }
 
+    const legacy = {
+      kubernetes_version: true,
+      weave_version: true,
+      rook_version: true,
+      contour_version: true,
+      registry_version: true,
```
Comment (emosbaugh) [inline web/src/installers/index.ts]: missing prometheus

**Pull request merged**